demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,732 @@
+        int numWritten = snprintf( _pDownloadDatas[reqIndex].rangeValueStr, sizeof( _pDownloadDatas[reqIndex].rangeValueStr ), "bytes=%d-%d", curByte, curByte + numReqBytes - 1 );
+        if( numWritten < 0 )
+        {
+            IotLogError( "Failed to write the header value: \"bytes=%d-%d\" . Error code: %d",
+                curByte,
+                curByte + numReqBytes - 1,
+                numWritten );
+            IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+        }
+
+        /* Set the user private data to use in the asynchronous callback context. */
+        _pDownloadDatas[reqIndex].reqNum = reqIndex;
+        _pDownloadDatas[reqIndex].currDownloaded = 0;
+        _pDownloadDatas[reqIndex].numReqBytes = numReqBytes;
+        asyncInfo.pPrivData = ( void* )( &(_pDownloadDatas[reqIndex]) );
+
+        /* Re-initialize the request to reuse the request. If we do not reinitialize then data from the last response
+           associated with this request will linger. */
+        httpsClientStatus = IotHttpsClient_InitializeRequest( &(_pReqHandles[reqIndex]), &(_pReqConfigs[reqIndex]) );
+        if( httpsClientStatus != IOT_HTTPS_OK )
+        {
+            IotLogError( "An error occurred in IotHttpsClient_InitializeRequest() with error code: %d", httpsClientStatus );
+            IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+        };
+
+        /* Send the request and receive the response asynchronously. This will schedule the async request. We 
+           will return immediately after scheduling. */
+        httpsClientStatus = IotHttpsClient_SendAsync( connHandle, _pReqHandles[reqIndex], &(_pRespHandles[reqIndex]), &( _pRespConfigs[reqIndex]) );
+        if( httpsClientStatus != IOT_HTTPS_OK )
+        {
+            IotLogError( "Failed to send the request asynchronously with error code: %d", httpsClientStatus );
+            /* If we failed to schedule an async request then this is an error and we should exit the loop. */
+            IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+        }
+        /* We increment by the number of request bytes. The S3 server should give us the range requested. If we
+           do not receive numReqBytes specified in the range, then we will have stop that async response processing in the
+           callbacks. */
+        curByte += numReqBytes;
+
+        /* If amount of file remaining to request is less than the current amount of bytes to request next time, then
+           udpdate the amount of bytes to request, on the next iteration, to be the amount remaining. */
+        if( ( _fileSize - curByte ) < numReqBytes )
+        {
+            numReqBytes = _fileSize - curByte;
+        }
+    }
+
+    /* If the status of the HTTPS Client library API is not OK after exiting the loop or if we timeout waiting for 
+       file to finish downloading, then we failed the demo. */
+    if(IotSemaphore_TimedWait( &( _fileFinishedSem ), IOT_HTTPS_DEMO_ASYNC_TIMEOUT_MS ) == false)
```

Comment:
@jainamt  As I understand this, the number of outstanding requests depends upon whether the file size is bigger than buffer (IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE ) or not.  If more than three requests are required then then code will wait (line 651).   Semwait is for all requests to finish. But  since semaphore initialized to 0 with max value 1,  if there are more than one outstanding request, Semwait will exit with success, which is probably not desired. Is that correct ?

Reply:
Yes we queue up three requests max and wait for them to finish. If the file requested is less than IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE * 3, the we will queue less requests.

This semaphore is posted to ONLY by the request that that downloads the final range.
https://github.com/aws/amazon-freertos/blob/feat/http_dev/demos/https/iot_demo_https_s3_download_async.c#L209